### PR TITLE
Add async login and logout by MetaMask

### PR DESCRIPTION
**Pull request by saneery**
#1709
**Merge after [PR#2215](https://github.com/poanetwork/blockscout/pull/2215)**

Add async login and logout by MetaMask

## Changelog

* Add web3.js in dependencies
* Login by MetaMask
* Async update top navbar with a user data on change account or logout

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/templates/stakes/_stakes_top.html.eex [hunk 1]
```diff
@@ -4,13 +4,19 @@
...
-      <%= render BlockScoutWeb.StakesView, "_stakes_stats_item_account.html", logged_in: @logged_in %>
+      <%= render BlockScoutWeb.StakesView, "_stakes_stats_item_account.html", logged_in: @logged_in, user: @user %>
       <!-- Buttons -->
       <div class="stakes-top-buttons">
-        <%= render BlockScoutWeb.CommonComponentsView, "_btn_add_full.html", text: gettext("Become Candidate"), extra_class: "js-become-candidate" %>
-        <%= render BlockScoutWeb.StakesView, "_stakes_btn_remove_pool.html", text: gettext("Remove My Pool"), extra_class: "js-remove-pool" %>
+        <%= if not @logged_in or not @user.has_pool do %>
+          <%= render BlockScoutWeb.CommonComponentsView, "_btn_add_full.html", text: gettext("Become a Candidate"), extra_class: "js-become-candidate" %>
+        <% end %>
+        <%= if @logged_in and @user.has_pool do %>
```
[the remark this answers is not in the document]
Comment (saneery) [reply apps/block_scout_web/lib/block_scout_web/templates/stakes/_stakes_top.html.eex]: @goodsoft Updated

File: apps/indexer/config/config.exs [hunk 2]
```diff
@@ -37,8 +37,11 @@ config :indexer,
   first_block: System.get_env("FIRST_BLOCK") || "0"
 
 # config :indexer, Indexer.Fetcher.ReplacedTransaction.Supervisor, disabled?: true
-# config :indexer, Indexer.Fetcher.BlockReward.Supervisor, disabled?: true
-config :indexer, Indexer.Fetcher.StakingPools.Supervisor, disabled?: true
+if System.get_env("POS_STAKING_CONTRACT") && System.get_env("POS_VALIDATORS_CONTRACT") do
+  config :indexer, Indexer.Fetcher.BlockReward.Supervisor, disabled?: true
+else
+  config :indexer, Indexer.Fetcher.StakingPools.Supervisor, disabled?: true
+end
```
Comment (goodsoft) [inline apps/indexer/config/config.exs]: This change does not belong in this PR at all.

File: apps/block_scout_web/lib/block_scout_web/templates/pools/index.html.eex [hunk 3]
```diff
@@ -6,10 +6,13 @@
...
     epoch_end_in: @epoch_end_in,
     block_number: @block_number,
-    logged_in: false
+    user: @user,
+    logged_in: @logged_in
   %>
 </div>
-<section data-page="stakes" class="container">
+<section data-page="stakes" class="container" 
+  data-user-address="<%= if @user, do: @user.address %>"
```
Comment (goodsoft) [inline apps/block_scout_web/lib/block_scout_web/templates/pools/index.html.eex]: Just a nitpick: I would place this attribute onto `.stakes-top-stats-item-address` element to encapsulate logging in logic there.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+export function reducer (state = initialState, action) {
+  switch (action.type) {
+    case 'WEB3_DETECTED': {
+      return Object.assign({}, state, { web3: action.web3 })
+    }
+    case 'AUTHORIZED': {
+      const a = state.account || null
+      if ((a !== action.account)) {
```
Comment (goodsoft) [inline apps/block_scout_web/assets/js/pages/stakes.js]: This looks a bit weird: a one-letter variable used only once, and double parentheses.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
+import $ from 'jquery'
+import { createStore, connectElements } from '../lib/redux_helpers.js'
+import Web3 from 'web3'
+
+export const initialState = {
+  web3: null
```
Comment (goodsoft) [inline apps/block_scout_web/assets/js/pages/stakes.js]: I would expect `address`, `user` be defined here as well.

Comment (goodsoft) [reply apps/block_scout_web/assets/js/pages/stakes.js]: Oh, and `account` too.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    case 'AUTHORIZED': {
+      const a = state.account || null
+      if ((a !== action.account)) {
+        $.getJSON('/set_session', { address: action.account })
+          .done(response => {
+            if (response.success === true) {
+              store.dispatch({ type: 'GET_USER' })
+            }
+          })
+      }
```
Comment (goodsoft) [inline apps/block_scout_web/assets/js/pages/stakes.js]: I think we should use a websocket for this.
Moreover, I don't see any reason to not return the account immediately without another `/delegator` round-trip.

File: apps/explorer/lib/explorer/chain.ex [hunk 7]
```diff
@@ -3183,6 +3184,22 @@ defmodule Explorer.Chain do
...
+  def delegator_info(address) do
+    query =
+      from(
+        address in Address,
+        where: address.hash == ^address,
+        left_join: delegator in StakingPoolsDelegator,
+        on: delegator.delegator_address_hash == address.hash and delegator.is_active,
+        left_join: pool in StakingPool,
+        on: pool.staking_address_hash == address.hash and pool.is_active,
+        group_by: address.hash,
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/chain.ex]: Isn't this `group_by` useless?

File: apps/explorer/lib/explorer/chain.ex [hunk 8]
```diff
@@ -3183,6 +3184,22 @@ defmodule Explorer.Chain do
...
+        select: [sum(delegator.stake_amount), sum(pool.self_staked_amount), count(pool) > 0]
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/chain.ex]: As this function is located in `Chain`, and it most likely will be reused, I would `select` a map here instead of a list of arbitrary unnamed values.

File: apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex [hunk 9]
```diff
@@ -22,6 +22,58 @@ defmodule BlockScoutWeb.PoolsController do
...
+      |> get_session(:address_hash)
+      |> delegator_info()
+
+    options = [
+      epoch_number: epoch_number,
+      epoch_end_in: epoch_end_block - block_number,
+      block_number: block_number,
+      user: user,
+      logged_in: user != nil
+    ]
```
Comment (goodsoft) [inline apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex]: A lot of code duplication, please make a helper function to generate this particular set of options.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      if (state.user === oldState.user) return
+      $.getJSON(state.accountPath, {type: 'JSON', template: 'top'})
+        .done(response => {
+          $el.html(response.content)
+          if (!state.user && state.web3) {
+            $('[data-selector="login-button"]').on('click', loginByMetamask)
+          }
+          if (!state.web3) {
+            $('[data-selector="login-button"]').on('click', redirectToMetamask)
+          }
```
Comment (Heimdell) [inline apps/block_scout_web/assets/js/pages/stakes.js]: These two `if`s above are better to be remade into
```
if (!state.user) {
  if (state.web3) {
    ...
  } else {
    ...
  }
}
```
Which will make intent and flow more clear.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+function getWeb3 () {
+  if (window.ethereum) {
+    let web3 = new Web3(window.ethereum)
+    console.log('Injected web3 detected.')
+
+    setInterval(function () {
+      web3.eth.getAccounts()
+        .then(accounts => {
+          var defaultAccount = accounts[0] || null
```
Comment (Heimdell) [inline apps/block_scout_web/assets/js/pages/stakes.js]: Since `null == undefined` (see [this js equality table](https://dorey.github.io/JavaScript-Equality-Table/)), you don't need `|| null`. You can check later if `x == null` and it will catch both cases.

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 6 above)
Comment (Heimdell) [reply apps/block_scout_web/assets/js/pages/stakes.js]: Is there any business-level difference between `AUTHORIZED` and `GET_USER` and `UPDATE_USER` states? (latter 2 don't even sound as _states_)

If not, what about making all this dispatch chain 1 async function that ends with...

```
store.dispatch({
  type: 'UPDATE_USER',
  user: response.delegator
})
```

This function also should be invoked strictly outside the reducer, since the reducer should only change state. It should be invoked from button.

All this is true, if we don't have business-level requirements to draw special things _while_ we logging in. However, reducer _shouldn't_ request `dispatch` by itself.

Comment (Heimdell) [reply apps/block_scout_web/assets/js/pages/stakes.js]: Also, how about creating `update_user(delegator)`-alike functions for each action in the program? That will save you some debug time when you try to `UPDATE_USR`.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+  switch (action.type) {
+    case 'WEB3_DETECTED': {
+      return Object.assign({}, state, { web3: action.web3 })
+    }
+    case 'AUTHORIZED': {
+      const a = state.account || null
+      if ((a !== action.account)) {
+        $.getJSON('/set_session', { address: action.account })
+          .done(response => {
+            if (response.success === true) {
```
Comment (Heimdell) [inline apps/block_scout_web/assets/js/pages/stakes.js]: Why? This looks unnecessary.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    },
+    render ($el, state, oldState) {
+      if (oldState.web3 === state.web3) return
+      if (state.web3) {
+        $el.unbind('click')
+        $el.on('click', loginByMetamask)
+      } else {
+        $el.unbind('click')
+        $el.on('click', redirectToMetamask)
+      }
```
Comment (Heimdell) [inline apps/block_scout_web/assets/js/pages/stakes.js]: This `if` above can be replaced with these 2 lines:

```
$el.unbind('click')
$el.on('click', state.web3 ?
    loginByMetamask
  : redirectToMetamask
)
```

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 10 above)
Comment (Heimdell) [reply apps/block_scout_web/assets/js/pages/stakes.js]: Other that that, `if (state.web3)` can be made into a `?:` invocation. If you don't like long exprs, a variable can be created.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+    setInterval(function () {
+      web3.eth.getAccounts()
+        .then(accounts => {
+          var defaultAccount = accounts[0] || null
+          if (defaultAccount !== store.getState().account) {
+            store.dispatch({ type: 'AUTHORIZED', account: defaultAccount })
+          }
+        })
+    }, 100)
```
Comment (Heimdell) [inline apps/block_scout_web/assets/js/pages/stakes.js]: What is the logic of this? Each 100ms you do `web.eth.getAccounts()` and then send `AUTHORISED` if the account is different with the one in the store?

Maybe, this should be done only once, via `setImmediate`/`setTimeout`?

Even aside from _reading store outside of reducer/renderer_, this code looks strange.

File: apps/explorer/lib/explorer/chain.ex (hunk 7 above)
Comment (Heimdell) [reply apps/explorer/lib/explorer/chain.ex]: Since `hash` here should be different for each row, `group_by` will do nothing useful - if will create a list of groups of size 1 each, so it can be dropped.

File: apps/block_scout_web/lib/block_scout_web/templates/stakes/_stakes_stats_item_account.html.eex [hunk 15]
```diff
@@ -1,33 +1,36 @@
...
-        %> POA20</span>
-        </span>
+    <span class="stakes-top-stats-label-item"><%= gettext "Staked" %>:
+      <%=
+        if @user[:staked] do
+          format_wei_value(@user.staked, :ether)
+        else
+          "-"
+        end
+      %>
```
Comment (Heimdell) [inline apps/block_scout_web/lib/block_scout_web/templates/stakes/_stakes_stats_item_account.html.eex]: These two `if`-blocks look same; what about making a helper?

**Comment by Heimdell** [comment]
A constructive advice about redux app: move all query/login logic into one function `startRelogging()`, which has local buffer for "last logged user" and inside `setImmediate(..., 1000)` asks `/set_session` and `/delegator` (delegator what?) from a backend.

File: apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex [hunk 16]
```diff
@@ -79,18 +131,68 @@ defmodule BlockScoutWeb.PoolsController do
...
+
+  defp get_token_balance(address, token_name) do
+    {:ok, balance} =
+      address
+      |> Chain.address_tokens_with_balance()
+      |> Enum.find_value(Wei.cast(0), fn token ->
+        if token.name == token_name do
+          Wei.cast(token.balance)
+        end
+      end)
```
Comment (goodsoft) [inline apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex]: Can't we select the needed balance from database immediately, instead of fetching all of them and filtering afterwards?

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 14 above)
Comment (saneery) [reply apps/block_scout_web/assets/js/pages/stakes.js]: I use `setInterval` because of a user may change selected account in metamask and I need to check account changes.
https://github.com/MetaMask/faq/blob/master/DEVELOPERS.md#ear-listening-for-selected-account-changes

**Comment by saneery** [comment]
@goodsoft @Heimdell I've updated the PR, could you look at this, please?

File: apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex (hunk 16 above)
Comment (saneery) [reply apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex]: I can create a function that takes a token name and returns the balance of this token

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 14 above)
Comment (goodsoft) [reply apps/block_scout_web/assets/js/pages/stakes.js]: > we somewhat reluctantly recommend

:100:

Comment (saneery) [reply apps/block_scout_web/assets/js/pages/stakes.js]: What I should use to check changes?

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+    store.dispatch({ type: 'WEB3_DETECTED', web3: web3 })
+
+    const sessionAcc = $('.stakes-top-stats-item-address').data('user-address')
+    if (sessionAcc) {
+      login(sessionAcc)
+    }
```
[the remark this answers is not in the document]
Comment (goodsoft) [reply apps/block_scout_web/assets/js/pages/stakes.js]: But really, I would avoid using session for this entirely, as server can't assume anything about user's identity.
We don't authenticate with server in any way, and all actions are performed using MetaMask API.

The only reason to store logged-in user in session is making sure that page is rendered with a correct account displayed and "Login with" button doesn't blink for a second before we authorize with MetaMask and do roundtrips to get delegator info.
We certainly can't believe to be _AUTHORIZED_ before that happens. In contrast, we should clear session and reset interface if for any reason we fail to authorize with MetaMask, i.e. user logs out or, more importantly, uninstalls plugin between visits to BlockScout website.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+    load ($el) {
+      return {
+        accountPath: $el.data('async-listing')
+      }
+    }
+  },
+  '[data-selector="stakes-top"]': {
+    render ($el, state, oldState) {
+      if (state.user === oldState.user) return
+      $.getJSON(state.accountPath, {type: 'JSON', template: 'top'})
```
Comment (goodsoft) [inline apps/block_scout_web/assets/js/pages/stakes.js]: Asynchronous network requests in a function named `render` set off an alarm in my head. :rotating_light: :rotating_light: :rotating_light: 
I'm not familiar with how Redux works, but that seems kinda wrong to me.

Also, I don't understand the purpose and mechanics of `accountPath` field in `state`.
It is not even defined in `initialState`.

File: apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex (hunk 16 above)
Comment (saneery) [reply apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex]: Well then it needs to be updated after we will add the fetching of token's contract address

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+    const sessionAcc = $('.stakes-top-stats-item-address').data('user-address')
+    if (sessionAcc) {
+      login(sessionAcc)
+    }
+  }
+}
+
+async function login (address) {
+  await $.getJSON('/set_session', { address: address })
+  let response = await $.getJSON('/delegator', { address: address })
```
[the remark this answers is not in the document]
Comment (saneery) [reply apps/block_scout_web/assets/js/pages/stakes.js]: I will join these requests, but I don't understand fo what we need to use websocket?

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (saneery) [reply apps/block_scout_web/assets/js/pages/stakes.js]: > fill it into state immediately and avoid a roundtrip (two, actually) to asynchronously get that info.
Why two?

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 18 above)
Comment (saneery) [reply apps/block_scout_web/assets/js/pages/stakes.js]: `accountPath` is gotten from `data-async-listing` attribute in the `pools/index.html.eex` template.
It's needed for pagination https://github.com/poanetwork/blockscout/blob/6702a8cdde86a3cfc575905726f93dd3a7f7139e/apps/block_scout_web/assets/js/lib/async_listing_load.js

I will rename it to `controllerPath` for clarity

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (goodsoft) [reply apps/block_scout_web/assets/js/pages/stakes.js]: GitHub doesn't allow me to reply in that conversation, so here it goes:

> I will join these requests, but I don't understand fo what we need to use websocket?

This is mostly a matter of convention:
- HTTP GET requests are used for fetching HTML resources, i.e. pages or their fully rendered fragments.
- `/api` scope exists for REST-based communication.
- Bidirectional communication on a loaded page is performed using socket channels.

These "rules" are not enforced in this project, though, but it's still better to stick them, so that system behavior is predictable and there's less cognitive load for other developers.

There are more important security concerns, though, that somehow slipped past my attention before.
You **must never** use GET requests for mutating requests (in this case, *modifying* session).
POST requests, on the other hand, are fine if and only if you employ CSRF protection techniques.

As dealing with CSRF adds complexity, while socket channel is there almost for free, it's just easier to use socket. ¯\\\_(ツ)\_/¯

File: apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex (hunk 16 above)
Comment (goodsoft) [reply apps/block_scout_web/lib/block_scout_web/controllers/pools_controller.ex]: Yeah, let's wait for #2313 then.

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 18 above)
Comment (goodsoft) [reply apps/block_scout_web/assets/js/pages/stakes.js]: Shouldn't we somehow use `createAsyncLoadStore` then? :thinking: 
Anyway, I can't get rid of the feeling that execution flow is terribly wrong here. @Heimdell, halp!

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (saneery) [reply apps/block_scout_web/assets/js/pages/stakes.js]: Oh, I got it

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 18 above)
Comment (saneery) [reply apps/block_scout_web/assets/js/pages/stakes.js]: > Shouldn't we somehow use createAsyncLoadStore then?

I use it in #2251 
https://github.com/poanetwork/blockscout/pull/2251/files#diff-34ea28cc2a7f417ac77a6f676ff7ed2dR167 for reloading pools list

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (saneery) [reply apps/block_scout_web/assets/js/pages/stakes.js]: What if I just render CSRF token and put it into redux store, and after use it to call POST methods?

[the remark this answers is not in the document]
Comment (goodsoft) [reply apps/block_scout_web/assets/js/pages/stakes.js]: I kindly ask you not to do that and just reuse the channel you create in #2221 :)

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+      $el.unbind('click')
+      $el.on('click', state.web3 ? loginByMetamask : redirectToMetamask)
+    }
+  },
+  '[data-async-load]': {
+    load ($el) {
+      return {
+        controllerPath: $el.data('async-listing')
+      }
+    }
```
Comment (Heimdell) [inline apps/block_scout_web/assets/js/pages/stakes.js]: I've noticed you `return` from `load()` here, but not in the previous element. Is that intended?

File: apps/block_scout_web/assets/js/pages/stakes.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (Heimdell) [reply apps/block_scout_web/assets/js/pages/stakes.js]: Plz don't "use" stuff from `store` to do or coordinate side-effects. The `reducer` must be a pure function that accumulates changes originating from events, so you can `render` it.

TL;DR: You can store CSRF token in store, but please do neither requests in `reducer` using in, nor read from the store outside of reducer.

File: apps/block_scout_web/assets/js/pages/stakes.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+  var win = window.open('https://metamask.io', '_blank')
+  win.focus()
+}
+
+async function loginByMetamask () {
+  try {
+    await window.ethereum.enable()
+    const accounts = await store.getState().web3.eth.getAccounts()
+
+    const defaultAccount = accounts[0] || null
```
Comment (Heimdell) [inline apps/block_scout_web/assets/js/pages/stakes.js]: Again, why `|| null`? The `null == undefined` is true, and both `null` and `undefined` are logical `false`.

**Comment by goodsoft** [comment]
Superseded by #2408
